pkg/grpc/universalapi/api_workflow.go

```diff
@@ -50,175 +47,196 @@ func (a *UniversalAPI) GetWorkflowAlpha1(ctx context.Context, in *runtimev1pb.Ge
+	if err != nil {
+		a.Logger.Debug(err)
+		return emptyResponse, err
 	}
-	return a.workflowActivity(ctx, in, method, messages.ErrResumeWorkflow)
-}
 
-// workflowActivity is a helper function to handle workflow requests for pause, resume, and terminate
-func (a *UniversalAPI) workflowActivity(ctx context.Context, in *runtimev1pb.WorkflowActivityRequest, method func(context.Context, *workflows.WorkflowReference) error, methodErr messages.APIError) (*runtimev1pb.WorkflowActivityResponse, error) {
-	if in.InstanceId == "" {
-		err := messages.ErrMissingOrEmptyInstance
+	req := &workflows.PauseRequest{
+		InstanceID: in.InstanceId,
+	}
+	if err := workflowComponent.Pause(ctx, req); err != nil {
+		err = messages.ErrPauseWorkflow.WithFormat(in.InstanceId, err)
 		a.Logger.Debug(err)
-		return &runtimev1pb.WorkflowActivityResponse{}, err
+		return emptyResponse, err
 	}
+	return emptyResponse, nil
+}
 
-	if in.WorkflowComponent == "" {
-		err := messages.ErrNoOrMissingWorkflowComponent
+// ResumeWorkflowAlpha1 is the API handler for resuming a workflow
+func (a *UniversalAPI) ResumeWorkflowAlpha1(ctx context.Context, in *runtimev1pb.ResumeWorkflowRequest) (*emptypb.Empty, error) {
+	emptyResponse := &emptypb.Empty{}
+	if err := a.validateInstanceId(in.InstanceId, false /* isCreate */); err != nil {
 		a.Logger.Debug(err)
-		return &runtimev1pb.WorkflowActivityResponse{}, err
+		return emptyResponse, err
 	}
 
-	if _, ok := a.CompStore.GetWorkflow(in.WorkflowComponent); !ok {
-		err := messages.ErrWorkflowComponentDoesNotExist.WithFormat(in.WorkflowComponent)
+	workflowComponent, err := a.getComponent(in.WorkflowComponent)
+	if err != nil {
 		a.Logger.Debug(err)
-		return &runtimev1pb.WorkflowActivityResponse{}, err
+		return emptyResponse, err
 	}
 
-	req := workflows.WorkflowReference{
+	req := &workflows.ResumeRequest{
 		InstanceID: in.InstanceId,
 	}
+	if err := workflowComponent.Resume(ctx, req); err != nil {
+		err = messages.ErrResumeWorkflow.WithFormat(in.InstanceId, err)
+		a.Logger.Debug(err)
+		return emptyResponse, err
+	}
+	return emptyResponse, nil
+}
 
-	err := method(ctx, &req)
-	if err != nil {
-		err = methodErr.WithFormat(in.InstanceId)
+func (a *UniversalAPI) validateInstanceId(instanceID string, isCreate bool) error {
+	if instanceID == "" {
+		return messages.ErrMissingOrEmptyInstance
+	}
+
+	if isCreate {
+		// Limit the length of the instance ID to avoid potential conflicts with state stores that have restrictive key limits.
+		maxInstanceIdLength := 64
+		if len(instanceID) > maxInstanceIdLength {
+			return messages.ErrInstanceIDTooLong.WithFormat(maxInstanceIdLength)
+		}
+
+		// Check to see if the instance ID contains invalid characters. Valid characters are letters, digits, dashes, and underscores.
+		// See https://github.com/dapr/dapr/issues/6156 for more context on why we check this.
+		for _, c := range instanceID {
+			if !unicode.IsLetter(c) && c != '_' && c != '-' && !unicode.IsDigit(c) {
```

Comment:
Does this need to support non-ASCII characters? I am not sure if that's a risk. The current code iterates through runes, and methods like `unicode.IsLetter` match the entire "letter" plane in the Unicode specs and not just ASCII letters.

```go
	var chr byte
// Must not use `for .. range` because that iterates through runes (i.e. UTF-8 characters), while a regular for loop iterates through bytes
	for i := 0; i < len(instanceID); i++ {
		chr = instanceID[i]
		if (chr < 'A' || chr > 'Z') && (chr < 'a' || chr > 'z') && (chr < '0' || chr > '9') && chr != '_' && chr != '-' {
			return messages.ErrInvalidInstanceID.WithFormat(instanceID)
		}
	}
```

Reply:
There's no requirement to support non-ASCII characters. I figured we can start out as being a bit strict and then loosen up the requirements over time based on user feedback.

Reply:
I agree. Can you change to use the loop as in my example above then? The current implementation does support unicode broadly

Reply:
Oh, I missed that detail. That's interesting because I actually started by borrowing logic from [token.IsIdentifier](https://github.com/golang/go/blob/598cf5e6ac1deb87f4931af195dad66f2b3f1672/src/go/token/token.go#L335) for this code. I'm not an expert on this topic by any means, but any idea why the go maintainers were okay with using `range` on a string instead of treating it as an array like you're suggesting?

Reply:
Resolving for now as I don't think this should be a blocker. I'd love to hear your opinion on this when you get back from vacation, though!